### PR TITLE
Refactor the tests for forecast.solar

### PR DESCRIPTION
**Pull request by klaasnicolaas**
This PR replaces the old tests with new ones, up to 94% codecoverage.

- Add new package: aresponses, syrupy and pytest-freezer
- Add new CI, which testing matrix for python 3.11, 3.12 and upload coverage to codecov (add key in repo secrets!)

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          cache: 'poetry'
+      - name: 🏗 Install workflow dependencies
+        run: |
+          poetry config virtualenvs.create true
+          poetry config virtualenvs.in-project true
+      - name: 🏗 Install dependencies
+        run: poetry install --no-interaction
+      - name: 🚀 Process coverage results
+        run: |
+          poetry run coverage combine coverage*/.coverage*
```
Comment (MartinHjelmare) [inline .github/workflows/tests.yaml:74]: Is this really needed? It's not possible to just upload all the files?

File: src/forecast_solar/forecast_solar.py [hunk 2]
```diff
@@ -85,14 +85,12 @@ async def _request(
         try:
             result = await dns.query("api.forecast.solar", "A")
         except DNSError as err:
-            raise ForecastSolarConnectionError(
-                "Error while resolving Forecast.Solar API address"
-            ) from err
+            msg = "Error while resolving Forecast.Solar API address"
+            raise ForecastSolarConnectionError(msg) from err
```
Comment (MartinHjelmare) [inline src/forecast_solar/forecast_solar.py:89]: Personally I don't like this  ruff rule. I'd disable it.

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (klaasnicolaas) [reply .github/workflows/tests.yaml:74]: I think it already makes 1 coverage file by default, so I could skip combining it indeed

File: src/forecast_solar/forecast_solar.py (hunk 2 above)
Comment (klaasnicolaas) [reply src/forecast_solar/forecast_solar.py:89]: I like that Ruff applies this rule, but it's a bit of personal preference 😅

Comment (MartinHjelmare) [reply src/forecast_solar/forecast_solar.py:89]: It's just an additional not needed local variable for the message. 🤷

Comment (klaasnicolaas) [reply src/forecast_solar/forecast_solar.py:89]: Will put it on the ignore list 👍🏻

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (klaasnicolaas) [reply .github/workflows/tests.yaml:74]: It ensures that both coverage artifacts (python 3.11 and 3.12) from the first job are combined into one file and then uploaded as xml to codecov.

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        run: |
+          poetry config virtualenvs.create true
+          poetry config virtualenvs.in-project true
+      - name: 🏗 Install dependencies
+        run: poetry install --no-interaction
+      - name: Lint with flake8
+        run: |
+          poetry run flake8 src/forecast_solar
+      - name: 🚀 Run pytest
+        run: poetry run pytest -v --cov src tests
```
Comment (MartinHjelmare) [inline .github/workflows/tests.yaml:41]:
```suggestion
        run: poetry run pytest -v --cov-report xml:coverage.xml --cov src tests
```

File: .github/workflows/tests.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      - name: ⬆️ Upload coverage artifact
+        uses: actions/upload-artifact@v4.3.3
+        with:
+          name: coverage-${{ matrix.python }}
+          path: .coverage
```
Comment (MartinHjelmare) [inline .github/workflows/tests.yaml:46]:
```suggestion
          path: coverage.xml
```

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (MartinHjelmare) [reply .github/workflows/tests.yaml:74]: We can remove this.

**Comment by klaasnicolaas** [comment]
My proposal is to keep the CI as I initially committed, because even with the last suggestions (one last commit) I see that the coverage file is not found. So I'd rather keep what I know for sure works, than continue to muddle through.

File: .github/workflows/tests.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+          python-version: ${{ env.DEFAULT_PYTHON }}
+          cache: 'poetry'
+      - name: 🏗 Install workflow dependencies
+        run: |
+          poetry config virtualenvs.create true
+          poetry config virtualenvs.in-project true
+      - name: 🏗 Install dependencies
+        run: poetry install --no-interaction
+      - name: 🚀 Process coverage results
+        run: poetry run coverage xml -i
```
Comment (MartinHjelmare) [inline .github/workflows/tests.yaml:73]: We don't need this step.

**Review by MartinHjelmare — COMMENTED** [review]
You can check this repo for a working workflow:
https://github.com/home-assistant-libs/zwave-js-server-python/

**Comment by klaasnicolaas** [comment]
zwave-js is using different upload/download artifact versions, 

Currently codecov does not see a config file, so nothing is uploaded...

**Comment by klaasnicolaas** [comment]
@MartinHjelmare can you add a codecov token secret to this repository?

File: pyproject.toml [hunk 6]
```diff
@@ -35,20 +35,23 @@ yarl = ">=1.6.0"
 Changelog = "https://github.com/home-assistant-libs/forecast_solar/releases"
 
 [tool.poetry.group.dev.dependencies]
+aresponses = "3.0.0"
 black = "24.4.2"
 covdefaults = "2.3.0"
 coverage = {version = "7.5.0", extras = ["toml"]}
```
Comment (MartinHjelmare) [inline pyproject.toml:41]: I don't think we need an explicit dependency on coverage anymore.
```suggestion
```

Comment (MartinHjelmare) [reply pyproject.toml:41]: Coverage is required by pytest-cov already.

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+          cache: 'poetry'
+      - name: 🏗 Install workflow dependencies
+        run: |
+          poetry config virtualenvs.create true
+          poetry config virtualenvs.in-project true
+      - name: 🏗 Install dependencies
+        run: poetry install --no-interaction
+      - name: 🚀 Upload coverage report
+        uses: codecov/codecov-action@v4.3.0
+        with:
```
Comment (MartinHjelmare) [inline .github/workflows/tests.yaml:74]:
```suggestion
        with:
          fail_ci_if_error: true
```

**Review by MartinHjelmare — APPROVED** [review]
Nice!

**Pull request merged**